xdcr.go

```diff
@@ -0,0 +1,245 @@
+
+	callback := func(event sgbucket.FeedEvent) bool {
+		docID := string(event.Key)
+		trace("Got event %s, opcode: %s", docID, event.Opcode)
+		dsName, err := r.getFromBucketCollectionName(event.CollectionID)
+		if err != nil {
+			warn("Could not find collection with ID %d for docID %s: %s", event.CollectionID, docID, err)
+			r.errorCount.Add(1)
+			return false
+
+		}
+
+		switch event.Opcode {
+		case sgbucket.FeedOpDeletion, sgbucket.FeedOpMutation:
+			if strings.HasPrefix(docID, sgbucket.SyncDocPrefix) && !strings.HasPrefix(docID, sgbucket.Att2Prefix) {
+				trace("Filtering doc %s", docID)
+				r.docsFiltered.Add(1)
+				return true
+
+			}
+
+			toDataStore, err := r.toBucket.NamedDataStore(dsName)
+			if err != nil {
+				warn("Replicating doc %s, could not find matching datastore for %s in target bucket", event.Key, dsName)
+				r.errorCount.Add(1)
+				return false
+			}
+
+			originalCas, err := toDataStore.Get(docID, nil)
+			if err != nil && !toDataStore.IsError(err, sgbucket.KeyNotFoundError) {
+				warn("Skipping replicating doc %s, could not perform a kv op get doc in toBucket: %s", event.Key, err)
+				r.errorCount.Add(1)
+				return false
+			}
+
+			/* full LWW conflict resolution is not implemented in rosmar yet
+
+			CBS algorithm is:
+
+			if (command.CAS > document.CAS)
+			  command succeeds
+			else if (command.CAS == document.CAS)
+			  // Check the RevSeqno
+			  if (command.RevSeqno > document.RevSeqno)
+			    command succeeds
+			  else if (command.RevSeqno == document.RevSeqno)
+			    // Check the expiry time
+			    if (command.Expiry > document.Expiry)
+			      command succeeds
+			    else if (command.Expiry == document.Expiry)
+			      // Finally check flags
+			      if (command.Flags < document.Flags)
+			        command succeeds
+
+
+			command fails
+
+			In the current state of rosmar:
+
+			1. all CAS values are unique.
+			2. RevSeqno is not implemented
+			3. Expiry is implemented and could be compared except all CAS values are unique.
+			4. Flags are not implemented
+
+			*/
+
+			if event.Cas <= originalCas {
+				trace("Skipping replicating doc %s, cas %d <= %d", docID, event.Cas, originalCas)
+				return true
+			}
+
+			toCollection, ok := toDataStore.(*Collection)
+			if !ok {
+				warn("Datastore is not of type Collection, is of type %T", toDataStore)
+				r.errorCount.Add(1)
+			}
+
+			err = writeDoc(ctx, toCollection, originalCas, event)
+			if err != nil {
+				warn("Replicating doc %s, could not write doc: %s", event.Key, err)
+				r.errorCount.Add(1)
+				return false
+
+			}
+			r.docsWritten.Add(1)
+		}
+
+		return true
+
+	}
+	return r.fromBucket.StartDCPFeed(ctx, args, callback, nil)
+}
+
+// Stop terminates the replication.
+func (r *XDCR) Stop(_ context.Context) error {
+	close(r.terminator)
+	r.terminator = nil
+	return nil
+}
+
+// writeDoc writes a document to the target datastore. This will not return an error on a CAS mismatch, but will return error on other types of write.
```

Comment:
I'm not sure what "other types of write" means in this comment.  Also, setWithMeta still needs to do a CAS check to ensure the document hasn't been mutated on the target between the time we evaluated LWW and the time we write, right?